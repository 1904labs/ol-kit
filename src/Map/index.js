export { default as Map } from './Map'
export * from './utils'
