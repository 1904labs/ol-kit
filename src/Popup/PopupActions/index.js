export * from './PopupActionCopyWkt'
export * from './PopupActionGoogleMaps'
export * from './PopupActionEdit'
