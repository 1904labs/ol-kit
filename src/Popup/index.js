export { default as Popup } from './Popup'
export { default as PopupBase } from './PopupBase'
export { default as PopupDefaultInsert } from './PopupInsert/PopupDefaultInsert'
