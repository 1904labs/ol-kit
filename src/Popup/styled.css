

.container {
  background-color: #FFFFFF;
  font-family: 'ArialMT', 'Arial';
  font-weight: 400;
  z-index: 99;
  box-shadow: 0px 1px 5px 0px rgba(0, 0, 0, 0.2),
  0px 2px 2px 0px rgba(0, 0, 0, 0.14),
  0px 3px 1px -2px rgba(0, 0, 0, 0.12);
}
/* position: ${p => p.inline ? 'relative' : 'absolute'};
display: ${p => p.show ? 'block' : 'none'};
opacity: ${p => p.transparent ? 0.8 : 1};
width: ${p => appendPx(p.width)};
height: ${p => appendPx(p.height)};
left: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).left}px;
top: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).top}px; */


.arrowBox::after, .arrowBox::before {
  border: solid transparent;
  content: " ";
  height: 0;
  width: 0;
  position: absolute;
  pointer-events: none;
}


/* ${props => props.position === 'click' ? 'right' : props.position}: 100%;
${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: 50%; */

.arrowBox::after {
  border-width: 16px;
}
/* margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -16px;
border-${props => props.position === 'click' ? 'right' : props.position}-color: ${props => props.position === 'bottom'
? '#ededed'
: '#ffffff'}; */

.arrowBox::before {
  border-width: 17px;
}
/* margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -17px;
border-${props => props.position === 'click' ? 'right' : props.position}-color: rgba(0, 0, 0, 0.12); */

.dragHandleIcon {
  height: 25px;
  width: 70px;
  cursor: move;
  padding: 3px;
  z-index: 100;
  position: absolute;
  margin: auto;
  top: 0;
  left: 0;
  right: 0;
}

/* color: ${props => props.color || `#fff`}; */

