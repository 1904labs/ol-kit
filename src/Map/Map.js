import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'
import debounce from 'lodash.debounce'
import { StyledMap } from './styled'
import { createMap, updateMapFromUrl, updateUrlFromMap } from './utils'
import ugh from 'ugh'
import en from 'locales/en'

// context should only be created when <Map> is mounted (see constructor), otherwise it's null so child comps don't use context
export let MapContext = null

/**
 * A Reactified ol.Map wrapper component
 * @component
 * @category Map
 * @since 0.1.0
 */
class Map extends React.Component {
  constructor (props) {
    super(props)

    // this is used to create a unique identifier for the map div
    this.target = `_ol_kit_map_${nanoid(6)}`

    // create a map context
    MapContext = React.createContext()
  }

  componentDidMount () {
    const { map, onMapInit, updateUrlDebounce, updateUrlFromView, updateViewFromUrl, urlViewParam } = this.props

    if (!map) {
      // if no map was passed, create the map
      this.map = createMap({ target: this.target })
      // force update to get this.map into the context
      this.forceUpdate()
    } else {
      this.map = map
    }

    if (updateUrlFromView) {
      const setUrl = () => updateUrlFromMap(this.map, urlViewParam)
      const mapMoveListener = debounce(setUrl, updateUrlDebounce)

      // update the url param after map movements
      this.map.on('moveend', mapMoveListener)
    }

    if (updateViewFromUrl) {
      // read the url to update the map from view info
      updateMapFromUrl(this.map, urlViewParam)
        .catch(ugh.error)
        .finally(() => onMapInit(this.map)) // always fire callback with map reference on success/failure
    } else {
      // callback that returns a reference to the created map
      onMapInit(this.map)
    }
  }

  getContextValue = () => {
    return {
      map: this.map,
      translations: en
    }
  }

  render () {
    const { children, fullScreen } = this.props

    return (
      <>
        <StyledMap
          id={this.target}
          fullScreen={fullScreen} />
        <MapContext.Provider value={this.getContextValue()}>
          {!this.map // wait for a map to exist before rendering children that need a ref to map
            ? null
            : children}
        </MapContext.Provider>
      </>
    )
  }
}

Map.defaultProps = {
  fullScreen: false,
  map: null,
  onMapInit: () => {},
  updateUrlDebounce: 400,
  updateUrlFromView: true,
  updateViewFromUrl: true,
  urlViewParam: 'view'
}

Map.propTypes = {
  /** any ol-kit children components will automatically be passed a reference to the map object via the `map` prop */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** if this is set to false, the map will fill it's parent container */
  fullScreen: PropTypes.bool,
  /** optionally pass a custom map */
  map: PropTypes.object,
  /** returns an initialized map object after optional animations complete */
  onMapInit: PropTypes.func,
  /** the length of debounce on map movements before the url gets updated */
  updateUrlDebounce: PropTypes.number,
  /** add map location coords + zoom level to url as query params */
  updateUrlFromView: PropTypes.bool,
  /** update map view based off the url param */
  updateViewFromUrl: PropTypes.bool,
  /** change the url param used to set the map location coords */
  urlViewParam: PropTypes.string
}

export default Map
