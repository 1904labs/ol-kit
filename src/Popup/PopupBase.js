import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Container, ArrowBox } from './styled'

/**
 * @component
 * @category Popup
 * @since 0.2.0
 */
class PopupBase extends Component {
  render () {
    const { show, width, height, inline, pixel, children, arrow } = this.props
    const arrowTranslator = {
      top: 'bottom',
      bottom: 'top',
      left: 'right',
      right: 'left'
    }

    return (
      <Container width={width} height={height} show={show} inline={inline} pixel={pixel} arrow={arrow}>
        <ArrowBox position={arrowTranslator[arrow]}>
          {children}
        </ArrowBox>
      </Container>
    )
  }
}

PopupBase.defaultProps = {
  arrow: 'none',
  height: 280,
  inline: false,
  pixel: [0, 0],
  show: false,
  width: 280
}

PopupBase.propTypes = {
  /** The position of the popup's arrow (`top`, `right`, `bottom`, `left` or `none`) */
  arrow: PropTypes.string,
  /** The content of the popup */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** The height of the popup */
  height: PropTypes.number,
  /** Render the component inline (without absolute positioning) */
  inline: PropTypes.bool,
  /** The pixel coordinate where the popup should render */
  pixel: PropTypes.array,
  /** Show/hide the popup */
  show: PropTypes.bool,
  /** The width of the popup */
  width: PropTypes.number
}

export default PopupBase
