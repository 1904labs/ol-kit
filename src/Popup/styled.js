import styled from 'styled-components'

const positionContainer = (arrowDirection, [x, y], width, height) => {
  if (arrowDirection === 'top') {
    return { top: y + 16, left: x - width / 2 }
  } else if (arrowDirection === 'right') {
    return { top: y - height / 2, left: x - width - 16 }
  } else if (arrowDirection === 'bottom') {
    return { top: y - height - 16, left: x - width / 2 }
  } else if (arrowDirection === 'left') {
    return { top: y - height / 2, left: x + 16 }
  } else if (arrowDirection === 'none') {
    return { top: y, left: x }
  }
}

export const Container = styled.div`
  display: ${p => p.show ? 'block' : 'none'};
  position: ${p => p.inline ? 'relative' : 'absolute'};
  background-color: #FFFFFF;
  font-family: 'ArialMT', 'Arial';
  font-weight: 400;
  z-index: 99;
  opacity: ${p => p.transparent ? 0.8 : 1};
  width: ${p => p.width}px;
  height: ${p => p.height}px;
  left: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).left}px;
  top: ${({ arrow, height, pixel, width }) => positionContainer(arrow, pixel, width, height).top}px;
  box-shadow: 0px 1px 5px 0px rgba(0, 0, 0, 0.2),
  0px 2px 2px 0px rgba(0, 0, 0, 0.14),
  0px 3px 1px -2px rgba(0, 0, 0, 0.12);
`

export const ArrowBox = styled.div`
  display: ${p => p.unset ? 'none' : 'block'};
  &::after, &::before {
    ${props => props.position === 'click' ? 'right' : props.position}: 100%;
    ${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &::after {
    border-${props => props.position === 'click' ? 'right' : props.position}-color: ${props => props.position === 'bottom'
  ? '#ededed'
  : '#ffffff'};
    border-width: 16px;
    margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -16px;
  }

  &::before {
    border-${props => props.position === 'click' ? 'right' : props.position}-color: rgba(0, 0, 0, 0.12);
    border-width: 17px;
    margin-${props => props.position === 'top' || props.position === 'bottom' ? 'left' : 'top'}: -17px;
  }
`
export const DragHandleIcon = styled.div`
  height: 25px;
  width: 70px;
  ${props => !props.inline && `left: 50vw`};
  ${props => props.inline && `transform: translateY(-100%); margin: auto`};
  cursor: pointer;
  color: ${props => props.color || `#fff`};
  outline:none;
  padding: 3px;
  z-index: 100;
  position: absolute;
  top: 30px;
  left: 0;
  right: 0;
`
