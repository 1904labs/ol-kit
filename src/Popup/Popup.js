import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import olInteractionDraw from 'ol/interaction/draw'

import { connectToMap } from 'Map'
import PopupBase from './PopupBase'
import PopupDefaultInsert from './PopupInsert/PopupDefaultInsert'
import { addMovementListener, getLayersAndFeaturesForEvent, getPopupPositionFromFeatures, removeMovementListener } from './utils'

/**
 * @component
 * @category Popup
 * @since 0.2.0
 */
class Popup extends Component {
  constructor (props) {
    super(props)

    this.state = {
      features: [],
      loading: false,
      popupPosition: {
        arrow: 'none',
        fits: false,
        pixel: [0, 0]
      },
      show: false
    }
  }

  componentDidMount () {
    const { map } = this.props

    // bind to map click events
    map.on('click', this.mapClickHandler)
  }

  componentWillUnmount () {
    const { map } = this.props

    map.un('click', this.mapClickHandler)
  }

  mapClickHandler = e => {
    // hide every map click and reshow if layers are detected
    this.hidePopup()

    // Get the interactions from the map as an array.
    const interactions = e.map.getInteractions().getArray()

    // This checks to see if there is an active Draw interaction on the map and prevents the popup showing if it returns true.
    if (interactions.find((i) => i instanceof olInteractionDraw && i.get('active'))) return

    this.checkForFeaturesAtClick(e)
  }

  checkForFeaturesAtClick = e => {
    const { map } = this.props
    /**
      this util returns an array of promises for each layer at the click location that resolve once wms features
      are converted to wfs and adds them to the map (immediately resolves for features already on the map)
    */
    const promises = getLayersAndFeaturesForEvent(e)

    // cancel if no layers at click location
    // do not add map movement listener or get features from layers
    if (!promises.length) return

    // when the map is panned, we need to re-calculate the position of the popup
    const popupMoveHandler = () => {
      const { show, features, popupPosition: lastPosition } = this.state
      const opts = { lastPosition } // use current position as lastPosition for animation when moving map

      // only compute new positions if the popup is showing
      if (show) {
        // this util returns a position object that is used by the popup component in render()
        const popupPosition = getPopupPositionFromFeatures({ map }, features, opts)

        this.setState({ popupPosition })
      }
    }

    // this adds three listeners (map move, resize, zoom), all of which should re-position the popup
    this.movementListener = addMovementListener(map, popupMoveHandler)
    // parse promises for layers and features
    this.getNewFeatures(e, promises)
  }

  getNewFeatures = async (e, promises) => {
    const { featuresFilter, onMapClick } = this.props // eslint-disable-line no-unused-vars

    // show popup in loading state while before resolving
    this.setState({ show: true, loading: true }, () => onMapClick(this.state))

    const layers = await Promise.all(promises)

    const parsedFeatures = layers.reduce((acc, { features, layer }) => {
      const layerFeatures = features.map(feature => {
        feature.set('_ol_kit_parent', layer)

        return feature
      })

      return [...acc, ...layerFeatures]
    }, [])

    // ol returns these in reverse z-index order
    const features = parsedFeatures.reverse()

    const popupPosition = getPopupPositionFromFeatures(e, features)

    this.setState({ features, loading: false, popupPosition }, () => onMapClick(this.state))
  }

  hidePopup = () => {
    const { map, onMapClick } = this.props
    const popupPosition = getPopupPositionFromFeatures({ map }, [])

    // stop tracking movement when popup show is set to false
    this.movementListener && removeMovementListener(this.movementListener)

    this.setState({ features: [], popupPosition, show: false }, () => onMapClick(this.state))
  }

  render () {
    const { children, map, show: propShow } = this.props
    const { features, loading, popupPosition: { arrow, pixel }, show: stateShow } = this.state
    const show = typeof propShow === 'boolean' ? propShow : stateShow // keep show prop as source of truth over state

    return (
      !show
        ? null
        : (
          ReactDOM.createPortal(
            <PopupBase pixel={pixel} arrow={arrow} {...this.props} show={show}>
              {children || ( // default ui if no children are passed
                <PopupDefaultInsert
                  features={features}
                  loading={loading}
                  onClose={this.hidePopup} />
              )}
            </PopupBase>,
            map.getTargetElement()
          )
        )
    )
  }
}

Popup.defaultProps = {
  featuresFilter: features => features,
  onMapClick: () => {},
  show: undefined
}

Popup.propTypes = {
  /** Pass components as children of Popup component */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** return a filtered array of features */
  featuresFilter: PropTypes.func,
  /** a reference to openlayers map object */
  map: PropTypes.object.isRequired,
  /** callback fired on map clicks with state object:
    {
      features: [],
      loading: false, // true after click before layers/features load
      popupPosition: {
        arrow: 'none',
        fits: true, // does the popup have room to render around the feature bbox
        pixel: [0, 0]
      },
      show: false // suggestion to display or hide popup -- <Popup show={sourceOfTruth}> (show prop takes priority)
    }
  */
  onMapClick: PropTypes.func,
  /** boolean that is respected over internal state */
  show: PropTypes.bool
}

export default connectToMap(Popup)
