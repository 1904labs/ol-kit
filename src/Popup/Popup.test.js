import React from 'react'
import { render, waitFor } from '@testing-library/react'
import Map from 'Map'
import Popup from './Popup'
import 'jest-styled-components'

describe('<Popup />', () => {
  it('should render default insert', async () => {
    const { container } = render(<Map><Popup /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(container.firstChild).toMatchSnapshot()
  })
})
