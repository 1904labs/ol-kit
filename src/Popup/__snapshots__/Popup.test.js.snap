// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> should render default insert 1`] = `
.c1 {
  position: absolute;
  width: 22px;
  bottom: 8px;
  z-index: 1;
  cursor: pointer;
  left: 14px;
}

.c0 {
  height: 100%;
  width: 100%;
  position: relative;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

<div
  class="c0"
  id="_ol_kit_map_test_id"
>
  <a
    class="c1"
    href="https://ol-kit.com/"
    target="_blank"
    title="Powered by ol-kit"
  >
    <svg
      viewBox="0 0 204.76 236.44"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <style>
          .cls-1
          {
          fill:#231f20
          }
        </style>
      </defs>
      <g
        id="Layer_2"
      >
        <path
          class="cls-1"
          d="M102.38 0L0 59.11v118.22l102.38 59.11 102.38-59.11V59.11zM106 15.64l87 50.26v102L80.49 104.09zM18 62.27l74.66-43.1-33.95 118L18 62.27zM43.8 189l-32-18.51V75.37l42.35 77.86zm10.49 6.06l7.82-27.17 25.12 46.19zm50.38 26.44l-37.93-69.67 10.4-36.13 106.4 60.3z"
          id="logo_examples"
        />
      </g>
    </svg>
  </a>
  <div
    class="ol-viewport"
    data-view="1"
    style="position: relative; overflow: hidden; width: 100%; height: 100%;"
  >
    <canvas
      class="ol-unselectable"
      style="width: 100%; height: 100%; display: block;"
    />
    <div
      class="ol-overlaycontainer"
    />
    <div
      class="ol-overlaycontainer-stopevent"
    />
  </div>
</div>
`;
